components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,256 @@
+use tikv_util::{self, set_panic_mark, worker::*};
+
+pub const DEFAULT_CHECK_INTERVAL: Duration = Duration::from_secs(10);
+
+pub struct RecoveryRunner {
+    db: Arc<DB>,
+    store_meta: Arc<Mutex<StoreMeta>>,
+    // Considering that files will not be too much, it is enough to use `Vec`.
+    damaged_files: Vec<FileInfo>,
+    max_hang_duration: Duration,
+    check_duration: Duration,
+}
+
+#[derive(Debug, Clone)]
+struct FileInfo {
+    // Corrupted file name. example: /000033.sst
+    name: String,
+    smallest_key: Vec<u8>,
+    largest_key: Vec<u8>,
+    overlap_region_ids: Vec<u64>,
+    // Time to generate recovery task, be used to record whether the timeout
+    start_time: Instant,
+}
+
+impl Runnable for RecoveryRunner {
+    type Task = String;
+
+    fn run(&mut self, sst: String) {
+        self.generate_scheduling_tasks(&sst);
+    }
+}
+
+impl RunnableWithTimer for RecoveryRunner {
+    fn get_interval(&self) -> Duration {
+        self.check_duration
+    }
+
+    fn on_timeout(&mut self) {
+        self.check_damaged_files();
+    }
+}
+
+impl RecoveryRunner {
+    pub fn new(
+        db: Arc<DB>,
+        store_meta: Arc<Mutex<StoreMeta>>,
+        max_hang_duration: Duration,
+        check_duration: Duration,
+    ) -> Self {
+        RecoveryRunner {
+            db,
+            store_meta,
+            damaged_files: vec![],
+            max_hang_duration,
+            check_duration,
+        }
+    }
+
+    fn generate_scheduling_tasks(&mut self, path: &str) {
+        if self.exist_scheduling_regions(path) {
+            return;
+        }
+
+        let live_files = self.db.get_live_files();
+        for i in 0..live_files.get_files_count() {
+            if path == live_files.get_name(i as i32) {
+                let f = FileInfo {
+                    name: live_files.get_name(i as i32),
+                    smallest_key: live_files.get_smallestkey(i as i32).to_owned(),
+                    largest_key: live_files.get_largestkey(i as i32).to_owned(),
+                    overlap_region_ids: Vec::new(),
+                    start_time: Instant::now(),
+                };
+
+                // only support recovering data key for now.
+                if !keys::validate_data_key(&f.smallest_key)
+                    || (!keys::validate_data_key(&f.largest_key)
+                        && f.largest_key.as_slice() != keys::DATA_MAX_KEY)
+                {
+                    set_panic_mark_and_panic(
+                        path,
+                        &format!(
+                            "key range mismatch, smallest key:{:?}, largest key:{:?}",
+                            &f.smallest_key, &f.largest_key
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                if !self.check_overlap_damaged_regions(f) {
+                    set_panic_mark_and_panic(path, "Can't find overlap regions");
```

Comment:
it maybe possible

Reply:
fine,  none overlap regions existed seems that delete sst operations can be executed directly